### ISSUE TITLE
[API Proposal]: Structural Delegates

### ISSUE THREAD

**Issue by timcassell** [dotnet/runtime#106582]
### Background and motivation

Motivating csharplang discussions: https://github.com/dotnet/csharplang/discussions/8343 https://github.com/dotnet/csharplang/discussions/7785

Today, delegates are nominal types, and delegates with matching signatures are not interchangeable. `Func<T, bool>` != `Predicate<T>`. It would be ideal to be able to declare delegates by their signature rather than by their nominal type, similar to how function pointers can be declared. `delegate<T, bool>` (or whatever syntax C# decides to go with).

Moreover, at a fundamental level, delegates need not be more than an object reference + a function pointer. The existing class-based delegates have a lot more baggage, making them less efficient than they theoretically could be.

### API Proposal

The runtime adds new structural delegate types, along with new IL metadata to use them.

A delegate declared in assembly A `delegate<T, bool>` and a delegate declared in assembly B `delegate<T, bool>` are the exact same type, and therefore are identity convertible. `delegate<ref T, void>` can be used also.

`System.Delegate` type already exists, so a new base type can be added.

```cs
namespace System;

public abstract class StructuralDelegate : ValueType
{
}
```

Yes, structural delegates are struct-based instead of class-based. It essentially looks like this:

```cs
public struct delegate<T, TResult> : IEquatable<delegate<T, TResult>>
{
    private object _obj;
    private void* _mPtr;
    
    // ...
}
```

These new structural delegates do not support Begin/EndInvoke or multi-cast like existing delegates.

To avoid possible tearing of the struct due to thread races, each read/write of a structural delegate will be made atomic by the runtime (this may be blocked by #105054 or #31911). If a struct contains structural delegate fields (or nested struct with structural delegate fields), copying that struct requires each structural delegate field to be copied atomically (the entire struct copy need not be atomic). As an optimization, the atomic r/w could be relaxed if the target is on the stack instead of the heap.

Additionally, to ensure that these structural delegate types are "safe", the GC should be taught to specially treat the function pointer as an ALC reference.

These types should also be specially treated similar to `Nullable<T>`, such that comparing to `null` becomes the same as comparing the pointer to zero, and boxing a default value results in a null object. With that in mind, `Nullable<delegate<>>` may also need to be treated specially.

---

There could also be additional APIs to convert to/from existing delegates, or bypassing the atomic r/w with Unsafe, and other APIs that exist for current delegates, but those can be left to future proposals if this ever goes anywhere.

### API Usage

```cs
public class MyEvent<T>
{
    // Not thread-safe, but this should get the point across.
    private HashSet<delegate<T, void>> _callbacks = new();

    public void AddListener(delegate<T, void> callback)
        => _callbacks.Add(callback);
    
    public void RemoveListener(delegate<T, void> callback)
        => _callbacks.Remove(callback);
    
    public void Raise(T arg)
    {
        foreach (var callback in _callbacks.ToArray())
        {
            callback(arg);
        }
    }
}
```

```cs
// Very common pattern
public class MyComponent : IDisposable
{
    private readonly MyEvent<T> _event;
    
    public MyComponent(MyEvent<T> e)
    {
        _event = e;
        // Existing delegate allocates here, new structural delegate does not allocate.
        e.AddListener(OnEvent);
    }
    
    public void Dispose()
    {
        // Existing delegate allocates here, new structural delegate does not allocate.
        _event.RemoveListener(OnEvent);
    }
    
    private void OnEvent(T arg)
    {
        // ...
    }
}
```

### Alternative Designs

Structural delegates are class-based like existing delegates. This avoids the issue with atomic r/w and ALC reference, but also adds allocations and other overhead.

### Risks

Enforcing atomic r/w could impact performance of shared generics in AOT runtimes.

**Comment by huoyaoyuan** [comment]
This would be a massive feature request, rather than API request.

I'm positive to this, but I'd hope we can include more if we are making changes to this.

**Comment by teo-tsirpanis** [comment]
I am not sure that the need for this feature justifies the required work to design and implement it. The request already mentions metadata changes and three quite major special cases for the JIT, the GC and the type system. With function pointers being always available for the performance-minded people that don't care about capturing objects or unloadability, I cannot see what niche struct delegates would fill.

Some of the ideas described here can be extracted into individual feature suggestions:

* Treating different delegate types with the same signature (such as `Func<T,bool>` and `Predicate<T>`) as the same types in the eyes of the compiler and the runtime.
* Adding an `()` operator like C++ has, or supporting directly invoking values of types with an `Invoke` method.
* Allowing delegate type definitions to omit the `Begin`/`EndInvoke` methods.
  * Isn't this already allowed?

> These new structural delegates do not support […] multi-cast like existing delegates.

I remember reading that multicast delegates are pay-to-play, i.e. if you don't use them they won't affect performance.

**Comment by neon-sunset** [comment]
I think the way for successful structural delegates is a generalized shape that can be either duck-type targeted or an interface abstraction (IDelegate, IAction, IFunc, etc.) that users can provide struct implementations for. Then Roslyn could target that by emitting anonymous struct-based delegate implementations for methods eligible for those. To better enable scenarios that are currently replicated by hand e.g in TensorPrimitives. Notably, function pointers are not a performance optimization in such case and do not enable monorphization and subsequent inlining of delegate calls.

**Comment by teo-tsirpanis** [comment]
> I think the way for successful structural delegates is a generalized shape provided by an interface abstraction (IDelegate, IAction, IFunc, etc.) that users can provide struct implementations for and constrain generic arguments on. Then Roslyn could target that by emitting anonymous struct-based delegate implementations for methods eligible for those. To better enable scenarios that are currently replicated by hand e.g in TensorPrimitives.

I understand that static genericity on functions instead of data types is sometimes needed, but I am not sure if it's a good idea to make this easily accessible by elevating it to a first-class language feature. I am concerned by its downsides (increase in generated machine code size and fragility when used across logical modules) and its potential for abuse which I have seen happening in C++.

> enable monorphization and subsequent inlining of delegate calls. Nor they enable struct or ref struct closures.

How common do you believe are the cases where this would be beneficial? I don't personally believe that merely low-level code like `TensorPrimitives` would meet the bar.

**Comment by timcassell** [comment]
> This would be a massive feature request, rather than API request.

Very true. I just worked from the issue template.

> I'd hope we can include more if we are making changes to this.

Can you elaborate?

**Comment by timcassell** [comment]
> I am not sure that the need for this feature justifies the required work to design and implement it.

Yeah, I kinda figured that would be the reaction to this proposal.

> I cannot see what niche struct delegates would fill.

It's a similar niche to ref structs, possibly wider reaching, since just about everyone uses delegates, while not everyone uses ref structs.

Considering the amount of effort put into improving performance of safe managed code in .Net recently, this proposal is along the same trajectory, and shouldn't be dismissed just because it's a large amount of work, IMO.

> * Treating different delegate types with the same signature (such as `Func<T,bool>` and `Predicate<T>`) as the same types in the eyes of the compiler and the runtime.

That looks like an issue that has been open for 9 years with no action. #4331
I don't know all the details about it, but I wouldn't be surprised if there are complications and possibly breaking changes by changing the type compatibility of existing delegates.

This proposal is a fresh start from scratch, new types, no worries about breaking anything.

> * Adding an `()` operator like C++ has, or supporting directly invoking values of types with an `Invoke` method.

I'm not sure where that came from. This proposal has absolutely nothing to do with an invoke operator.

> * Allowing delegate type definitions to omit the `Begin`/`EndInvoke` methods.
>   
>   * Isn't this already allowed?
> 
> > These new structural delegates do not support […] multi-cast like existing delegates.
> > I remember reading that multicast delegates are pay-to-play, i.e. if you don't use them they won't affect performance.

Those were mentioned, not because it may or may not be possible to exclude them on existing delegates, but just to highlight that the new types absolutely will _not_ support them.

**Comment by timcassell** [comment]
> I think the way for successful structural delegates is a generalized shape provided by an interface abstraction (IDelegate, IAction, IFunc, etc.) that users can provide struct implementations for and constrain generic arguments on. Then Roslyn could target that by emitting anonymous struct-based delegate implementations for methods eligible for those. To better enable scenarios that are currently replicated by hand e.g in TensorPrimitives.

That looks like the functional interfaces proposal. https://github.com/dotnet/csharplang/issues/3452

That is similar to this proposal, but serves a different purpose. Namely, it can achieve zero allocation by using variable-sized structs, while this is a fixed-size type.

**Comment by timcassell** [comment]
I think this would be the preferred delegate type to use in most situations, but I wouldn't say those would be obsolete. Tons of APIs exist today that use those delegate types, and those aren't going anywhere. If you want to use the multi-cast feature of delegates (C# `event`), and don't want to allocate a custom event class, you would use those types.

**Comment by neon-sunset** [comment]
> How common do you believe are the cases where this would be beneficial? I don't personally believe that merely low-level code like `TensorPrimitives` would meet the bar.

`TensorPrimitives` is simply the latest example of a general "zero-cost value delegate" case that numerous libraries need and have to reimplement today manually.

The lowest effort kind of improvement in this area comes down to simply introducing the aforementioned interfaces and teaching Roslyn that method invocation operator applies to types implementing them.

This can be subsequently augmented with Roslyn learning to generate anonymous struct and ref struct[^1] delegates with respective closure types, possibly offering better UX with generic arguments displayed as `Where<int, impl IFunc<int, bool>>`. Notably, the area of this change would belong to dotnet/csharplang and an API proposal to dotnet/runtime but would *not* require changes in the runtime itself, beyond API work in CoreLib.

Once done, this could open up the possibility to introduce value-delegate taking method overloads to `System.Linq` which would further cut down on LINQ overhead without introducing breaking changes to existing customers or them to make changes in order to gain performance. This is in spirit of similar changes like `params ReadOnlySpan<T>`.

[^1]: https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-13#ref-struct-interfaces

**Comment by huoyaoyuan** [comment]
Finally have time to leave my words here.

With stack promotion finally being a thing, it's now less important for being struct or class. If we don't provide extra support for reflection etc, structs of 2-pointers size would be OK.

I'm more interested about how can we provide implementation about this. For nominal types, the pain is that we can't declare one method that satisfies different arities. Without changing this, the implementation of structural delegates can only be runtime magic. If we have variable arity support, we may declare them in ordinal managed code. Hypothetical syntax:

```csharp
struct StructuralDelegate<..TArgs, TReturn>
    where TReturn : allows ref struct, allows void
{
    private object? target;
    private delegate*<object?, ..TArgs, TReturn> fnptr;
    public TReturn Invoke(..TArgs args) => fnptr(target, args);
}
```

**Comment by julealgon** [comment]
> With stack promotion finally being a thing, it's now less important for being struct or class.

@huoyaoyuan would you mind elaborating a bit on this or just providing some link with more information about it? It is the first time I hear about it so I'm curious.

**Comment by huoyaoyuan** [comment]
> > With stack promotion finally being a thing, it's now less important for being struct or class.
> 
> @huoyaoyuan would you mind elaborating a bit on this or just providing some link with more information about it? It is the first time I hear about it so I'm curious.

See more at #103361, and more follow-up PRs.

Note that I'm not saying that allocation is *not a problem*. It's just having different impact now. Large structs have their issues.
If we can implement structural delegates in 2 pointers and cut unnecessary features, then struct should be better.

**Comment by hez2010** [comment]
The idea is great, but I don't think such thing can be implemented in the proposed way. 
Existing generic metadata has too many restrictions like no support for variadic type parameters as well as `ref/in/out/void` parameters etc. I expect "structural delegates" to work like how function pointers work today, i.e., introducing new `TypeDesc` for structural delegates, and encoding structural delegate types directly in the signature.

**Comment by timcassell** [comment]
> The idea is great

Why the downvote then? 

> but I don't think such thing can be implemented in the proposed way. Existing generic metadata... 

I'm not sure what you mean. I proposed to add new IL metadata.

**Comment by timcassell** [comment]
> I'm more interested about how can we provide implementation about this. For nominal types, the pain is that we can't declare one method that satisfies different arities. Without changing this, the implementation of structural delegates can only be runtime magic. If we have variable arity support, we may declare them in ordinal managed code. Hypothetical syntax:
> 
> ```cs
> struct StructuralDelegate<..TArgs, TReturn>
>     where TReturn : allows ref struct, allows void
> {
>     private object? target;
>     private delegate*<object?, ..TArgs, TReturn> fnptr;
>     public TReturn Invoke(..TArgs args) => fnptr(target, args);
> }
> ```

I really like the idea, and I think it deserves its own proposal. I would love to be able to write a method that can return `ValueTuple<..TArgs>`.

**Comment by julealgon** [comment]
> I really like the idea, and I think it deserves its own proposal. I would love to be able to write a method that can return `ValueTuple<..TArgs>`.

I'm sure I'm missing something... I thought there would already be a proposal for variadic generics, but I searched both here and on `csharplang` and didn't find a single issue proposing it?

I would've assumed folks would be proposing that ever single generics was introduced 😅 

Maybe it just got stuck in discussions and was never promoted to an issue?

Would definitely be nice to finally stop needing to maintain dozens of variations of `Action`, `Func` and tuples, besides all other obvious benefits it would bring.

**Comment by tannergooding** [comment]
Variadics don't necessarily translate well to generics. Despite generics and templates having similar syntax, they are not the same thing and the implications of them are drastically different. This is particularly relevant as it pertains to the general ABI (Application Binary Interface) of an exposed API.

Outside of templates, variadic functions themselves necessitate a different calling convention and specialized support for parameters. There is no support for variadic returns on concrete methods even in C++. The reason variadic returns look like they exist for templates is because templates don't really exist, they're "erased" at C++ compilation time. That is, they're just symbols that get specialized based on their usage and synthesize types as part of compilation. You can't "export" them from a library, the non-specialized implementation needs to exist in the header itself, you can't do `RTTI` over some non specialized type, etc (there is no `typeof(List<>)` like .NET has). So from the ABI perspective, there is no `MyType<...TArgs>` return, there are many individual methods like `MyType<int>` or `MyType<int, int>`, etc  that happened to have used templates to simplify the implementation. -- They are more akin to the erased generics that Java does.

Versioning IL metadata is also incredibly breaking and we want to avoid it wherever possible; so if you really want a feature then the best approach is to find a way it can be represented without needing to go and break the entire ecosystem, such as by using attributes or other mechanisms that let the runtime and languages work with it using the existing support. In some cases there might not be an obvious alternative and it may require such revisions to IL, but that also makes it drastically less likely to happen in the near term.

**Comment by huoyaoyuan** [comment]
> Variadics don't necessarily translate well to generics. Despite generics and templates having similar syntax, they are not the same thing and the implications of them are drastically different. This is particularly relevant as it pertains to the general ABI (Application Binary Interface) of an exposed API.

To make it more clear: I'm not talking about va_args in C, or managed va_args in IL/managed C++. Instead, I'm requesting an general IL/C# syntax for "any signature". The signature is applicable to any calling convention because it will just pass every arguments and returns as-is. It's effectively using signatures as generic parameters.

It's a goal of higher-kind generic, and will naturally cover many runtime magics for delegates.

**Comment by julealgon** [comment]
@tannergooding I'm aware of the differences between C++ templates and the C# generics system. I wasn't trying to make any point about it being _easy_ or even _feasible_ for that matter to do variadic generics in C#, but just expressing surprise in not seeing a proposal case anywhere (not even a closed one). 

I think anyone would immediately frown upon seeing _this_ for the first time, and would immediately jump to such proposal:
![image](https://github.com/user-attachments/assets/2987071c-b451-45ad-9157-1c73fbb31dd7)

**Comment by tannergooding** [comment]
I was simply trying to give some suggestions on how to approach this in a way that would give the highest possibility of a solution being found.

If the community wants some variadic generics like support, then someone should start by writing up proposal which covers how it could integrate with existing IL metadata in a pay for play mechanism and some basics on how it could reasonably be handled in reflection, the type system, etc.

**Comment by timcassell** [comment]
> In some cases there might not be an obvious alternative and it may require such revisions to IL, but that also makes it drastically less likely to happen in the near term.

Tbh, I don't expect this feature to be implemented any time soon. Besides the IL revisions, there is also the atomic r/w of 2 pointers which has been requested for 8+ years, and GC tracking of a managed function pointer which is new and probably risky. 

This is a chance to learn from 22 years of imperfect delegates, and to do it right from the ground up. I wouldn't want to shoe horn it into existing functionalities just to get it done faster if it's not the best it can be.

**Comment by tannergooding** [comment]
We can't require atomic r/w of 2 pointers, its not guaranteed to exist.

It's one of the reasons we can't trivially expose a BCL API for it, although a hardware intrinsic per platform might be feasible.

**Comment by timcassell** [comment]
> We can't require atomic r/w of 2 pointers, its not guaranteed to exist.
> 
> It's one of the reasons we can't trivially expose a BCL API for it, although a hardware intrinsic per platform might be feasible.

It doesn't have to be an API only for 2 pointers. The `BitwiseAtomic<T>` [proposal](#105054) suggests to fall back to a SpinLock if the architecture doesn't support it.

**Comment by timcassell** [comment]
> I'm sure I'm missing something... I thought there would already be a proposal for variadic generics, but I searched both here and on `csharplang` and didn't find a single issue proposing it?

I found this old issue on roslyn repo. https://github.com/dotnet/roslyn/issues/5058 Not really any newer issues since then, and LDM would not likely adopt the proposed solution (hacky, compiler-driven). Probably more likely to gain traction as a runtime feature.

**Comment by nczsl** [comment]
I think that the delegate design and other related features are historical baggage and should be phased out as soon as possible. We should directly emulate the function pointers introduced in .NET 5. If `delegate*<T1...Tn>` exists, just follow the pattern. `delegate<T1...Tn>` should be used without hesitation; there shouldn't be so many concerns. Those historically proven inefficient and ineffective features in C# should be replaced as soon as possible. I want to emphasize that this is of great significance. Combined with tail recursion, existing C# support for lambda expressions, pattern matching, and switch expressions, this reform will immediately transform C# into a full-fledged FP paradigm language. Theoretically, business logic can be built entirely on expressions without writing any imperative statements. This is a crucial paradigm because it allows C# to better perform parallelization, thanks to FP's inherent concurrency friendliness! This is very important. Combined with C#'s well-established OOP paradigm and unsafe capabilities, a single language will possess the capabilities of all the most advanced languages ​​currently available, which is extremely useful. I want to emphasize that unsafe-clike, OOP, and FP are all very important and indispensable.

**Comment by huoyaoyuan** [comment]
> I think that the delegate design and other related features are historical baggage and should be phased out as soon as possible. We should directly emulate the function pointers introduced in .NET 5.

Delegates and function pointers are different things. Delegate has an additional closure.

> If `delegate*<T1...Tn>` exists, just follow the pattern. `delegate<T1...Tn>` should be used without hesitation; there shouldn't be so many concerns. Those historically proven inefficient and ineffective features in C# should be replaced as soon as possible.

Function pointers are *usually insufficient and being moved away from*. Even with the longest history of function pointers, C++ has introduced closures (`std::function`).

> this reform will immediately transform C# into a full-fledged FP paradigm language.

This is never the goal for C#. C# adapts FP patterns when FP patterns are better.

> Theoretically, business logic can be built entirely on expressions without writing any imperative statements. This is a crucial paradigm because it allows C# to better perform parallelization, thanks to FP's inherent concurrency friendliness!

Delegates are much more friendly to use than function pointers. Lack of closure make things a lot more complex.

This feature request is mostly about performance and higher-kind generic on delegate types. It's not related to make C# functional.

**Comment by nczsl** [comment]
来自 Outlook<http://aka.ms/weboutlook>  c++ 有闭包就有呗，关键是 我觉得C# 现在已经有了 delegate*<T1...Tn>这种定义，完全可以扩展一下加上 delegate<T1...Tn> 好吗？ 你不觉得Func<T...> 整一堆在系统 库里 并不太优雅吗？原因我之前已经说过了，委托或许有用，但它也有性能成本，而且委托在很多时候显得没有意义，它们在定义的时候重点是输入参数列表，和输出参数 这些信息，而不是多了一个完全没用的名称，你不觉得吗？只要 函数签名一样，那么不论你定义多少个不同名称的委托都没什么意义，看出来这个语法的问题没有？ 重点是 delegate<T1...Tn>  这种简洁直白的定义方式好吗？
________________________________
发件人: Huo Yaoyuan ***@***.***>
发送时间: 2025年11月11日 22:39
收件人: dotnet/runtime ***@***.***>
抄送: nczsl ***@***.***>; Comment ***@***.***>
主题: Re: [dotnet/runtime] [Feature Proposal]: Structural Delegates (Issue #106582)

[https://avatars.githubusercontent.com/u/5644458?s=20&v=4]huoyaoyuan left a comment (dotnet/runtime#106582)<https://github.com/dotnet/runtime/issues/106582#issuecomment-3517225592>

I think that the delegate design and other related features are historical baggage and should be phased out as soon as possible. We should directly emulate the function pointers introduced in .NET 5.

Delegates and function pointers are different things. Delegate has an additional closure.

If delegate*<T1...Tn> exists, just follow the pattern. delegate<T1...Tn> should be used without hesitation; there shouldn't be so many concerns. Those historically proven inefficient and ineffective features in C# should be replaced as soon as possible.

Function pointers are usually insufficient and being moved away from. Even with the longest history of function pointers, C++ has introduced closures (std::function).

this reform will immediately transform C# into a full-fledged FP paradigm language.

This is never the goal for C#. C# adapts FP patterns when FP patterns are better.

Theoretically, business logic can be built entirely on expressions without writing any imperative statements. This is a crucial paradigm because it allows C# to better perform parallelization, thanks to FP's inherent concurrency friendliness!

Delegates are much more friendly to use than function pointers. Lack of closure make things a lot more complex.

This feature request is mostly about performance and higher-kind generic on delegate types. It's not related to make C# functional.

―

**Comment by nczsl** [comment]
而且， 你想啊，我在使用的时候 void M<T>(delegate<T,T> x); 好用呢？还是 void M<T>( 这里还要用Func<T,T>) 这种不纯粹的东西好

来自 Outlook<http://aka.ms/weboutlook>
________________________________
发件人: mao gang ***@***.***>
发送时间: 2025年11月12日 9:23
收件人: dotnet/runtime ***@***.***>
主题: 回复: [dotnet/runtime] [Feature Proposal]: Structural Delegates (Issue #106582)



来自 Outlook<http://aka.ms/weboutlook>  c++ 有闭包就有呗，关键是 我觉得C# 现在已经有了 delegate*<T1...Tn>这种定义，完全可以扩展一下加上 delegate<T1...Tn> 好吗？ 你不觉得Func<T...> 整一堆在系统 库里 并不太优雅吗？原因我之前已经说过了，委托或许有用，但它也有性能成本，而且委托在很多时候显得没有意义，它们在定义的时候重点是输入参数列表，和输出参数 这些信息，而不是多了一个完全没用的名称，你不觉得吗？只要 函数签名一样，那么不论你定义多少个不同名称的委托都没什么意义，看出来这个语法的问题没有？ 重点是 delegate<T1...Tn>  这种简洁直白的定义方式好吗？
________________________________
发件人: Huo Yaoyuan ***@***.***>
发送时间: 2025年11月11日 22:39
收件人: dotnet/runtime ***@***.***>
抄送: nczsl ***@***.***>; Comment ***@***.***>
主题: Re: [dotnet/runtime] [Feature Proposal]: Structural Delegates (Issue #106582)

[https://avatars.githubusercontent.com/u/5644458?s=20&v=4]huoyaoyuan left a comment (dotnet/runtime#106582)<https://github.com/dotnet/runtime/issues/106582#issuecomment-3517225592>

I think that the delegate design and other related features are historical baggage and should be phased out as soon as possible. We should directly emulate the function pointers introduced in .NET 5.

Delegates and function pointers are different things. Delegate has an additional closure.

If delegate*<T1...Tn> exists, just follow the pattern. delegate<T1...Tn> should be used without hesitation; there shouldn't be so many concerns. Those historically proven inefficient and ineffective features in C# should be replaced as soon as possible.

Function pointers are usually insufficient and being moved away from. Even with the longest history of function pointers, C++ has introduced closures (std::function).

this reform will immediately transform C# into a full-fledged FP paradigm language.

This is never the goal for C#. C# adapts FP patterns when FP patterns are better.

Theoretically, business logic can be built entirely on expressions without writing any imperative statements. This is a crucial paradigm because it allows C# to better perform parallelization, thanks to FP's inherent concurrency friendliness!

Delegates are much more friendly to use than function pointers. Lack of closure make things a lot more complex.

This feature request is mostly about performance and higher-kind generic on delegate types. It's not related to make C# functional.

―

**Comment by huoyaoyuan** [comment]
> 我觉得C# 现在已经有了 delegate*<T1...Tn>这种定义，完全可以扩展一下加上 delegate<T1...Tn> 好吗？

`delegate*<T1...Tn>` 在 .NET Framework 1.0 的类型系统中**20年前早已存在**，只是C#最近才加上了支持。
`delegate<T1...Tn>` 在CLR类型系统中**从未存在**，两个签名一致的委托是不同的类型，无法转换。

**Comment by TahirAhmadov** [comment]
I'm confused as to why variadic type arguments entered the conversation. Isn't it orthogonal to this proposal? With `struct` delegates, we would still create new types, just like with existing delegates. If (it's a big if) variadic type arguments become a thing, they become available to both the existing delegates and these new struct delegates.
And I want to express my dislike for types like `Action<T0, ...>`, etc. - IMO this is abuse of the generic system. If you need a delegate which serves a particular need, just declare that delegate type, give it a name, and all the parameters it needs. Same with tuples - every time I used a tuple even in a type's private code (let alone public surface), it came back to bite me in some way, and I always change to `record struct` as a personal rule.
It's sad to see such a great proposal for getting rid of the delegate multicast legacy baggage, together with performance improvement of avoiding allocations for a massive number of use cases, to be hampered by orthogonal concerns - and concerns I'd argue are invalid to begin with (like the need to support `Action<T0, ...>` etc.)

**Comment by timcassell** [comment]
> I'm confused as to why variadic type arguments entered the conversation. Isn't it orthogonal to this proposal? With struct delegates, we would still create new types, just like with existing delegates. If (it's a big if) variadic type arguments become a thing, they become available to both the existing delegates and these new struct delegates.

I think the argument is, since this requires runtime magic anyway to support `<T1, ...Tn>`, general variadic support may as well be added while we're at it. It's not _really_ required, after all runtime magic works for `delegate*<T1, ...Tn>` since forever.

> If you need a delegate which serves a particular need, just declare that delegate type, give it a name, and all the parameters it needs.

Part of this proposal is about solving the problem of assignability of delegates with the same signature. If you have to declare all of your delegates nominally like we do today, it's not possible (https://github.com/dotnet/csharplang/discussions/7785 proposes just that).

**Comment by TahirAhmadov** [comment]
> Part of this proposal is about solving the problem of assignability of delegates with the same signature.

It becomes more philosophical, but assigning a delegate (class type or struct type) to another delegate which matches by signature is a strong code smell. A delegate is much more than just the parameter/return types; it's what is expected by calling it. Nominal types allow you set expectations correctly.

If it is truly needed, I mean there are always edge cases, then an explicit conversion can be devised. Currently, you can do `Func<int, bool> foo = ...; var bar = new Predicate<int>(foo);` - I'm sure something like this can be accomplished with struct delegates, perhaps using static interface/abstract members?
```cs
struct delegate bool Del1(int x);
// elsewhere, different project, doesn't mattter
struct delegate bool Del2(int x);

Del1 foo = ....;
var bar = Del2.ConvertFrom(foo);

// behind the scenes:
public abstract class StructuralDelegate : ValueType
{
  public static abstract Type[] ParameterTypes { get; } // 
  public static abstract Type ReturnType { get; }

  public static TTo Convert<TFrom, TTo>(TFrom source)
    where TFrom : struct, StructuralDelegate
    where TTo : struct, StructuralDelegate
  {
    if(!TFrom.ParameterTypes.SequenceEqual(TTo.ParameterTypes)) throw new Exception(); 
    // etc. validation

    return new TTo {
      _obj = source._obj, // here these will probably need to be abstract properties
      _mPtr = source._mPtr
    };
  }
}
struct Del2 : StructuralDelegate
{
  public static Del2 ConvertFrom<T>(T source) 
    where T : struct, StructuralDelegate
  {
    return Convert<T, Del2>(source);
  }
}
```
Basically the objective here is a safe, allocation-free (chicken scratch LINQ calls for validation notwithstanding) conversion system, which has to be invoked explicitly.

**Comment by timcassell** [comment]
I mean, it's an opinion. I personally despise nominal delegates. If we had structural delegates, I would just say if you want named delegates, use type aliasing.

**Comment by colejohnson66** [comment]
> It becomes more philosophical, but assigning a delegate (class type or struct type) to another delegate which matches by signature is a strong code smell.

Yes, and no.

Before the `Action` and `Func` zoo we have now, the .NET Framework team created one-off delegates that *are* indeed *functionally equivalent*. For example, `Predicate<T>` and `Func<T, bool>` are conceptually the same thing. Not to mention that, ignoring the object header, the delegate allocations between the two (meaning the target method and closure object) would be identical! But because they are two distinct types, if I have a `Func<T, bool>` and want to do `List<T>.Find`, I need to:

```cs
List<T> list = ...;
Func<T, bool> predicate = ...;
T? found = list.Find(x => predicate(x));
```

Ignoring the extra closure allocation there, it makes me wince. Thankfully, these situations are rare, but they do crop up if your predicate was handed to you from somewhere else (such as a parameter).

**Comment by TahirAhmadov** [comment]
@colejohnson66 well in the example you cited, the problem arose specifically because `Func<T, bool>` was created and became the type of choice for methods like `Where`. In my world, `Func<T, TReturn>` wouldn't exist in the first place, `Where` would expect `Predicate<T>`, and everything would be fine. Meaning, the solution created the problem, so to speak.

**Comment by TahirAhmadov** [comment]
More importantly, though, this debate can be decoupled from structural delegates entirely; it just complicates the discussion around the core issues here which have to do with the runtime, the "function pointer keeps method's assembly alive". Or maybe I am missing something?

On the topic of method reference, doesn't a `MethodInfo` reference already provide the needed behavior of keeping the assembly loaded? Couldn't the struct delegate keep a `MethodInfo` reference, and then JIT would simply optimize the call to `mi.Invoke(...)` as needed?

**Comment by timcassell** [comment]
> More importantly, though, this debate can be decoupled from structural delegates entirely; it just complicates the discussion around the core issues here which have to do with the runtime, the "function pointer keeps method's assembly alive". Or maybe I am missing something?

I don't think it should be decoupled, though. If struct delegates become a thing, they will either be structural or nominal, and likely never revisited again. I strongly favor structural, but it's a decision that must be made.

> On the topic of method reference, doesn't a `MethodInfo` reference already provide the needed behavior of keeping the assembly loaded? Couldn't the struct delegate keep a `MethodInfo` reference, and then JIT would simply optimize the call to `mi.Invoke(...)` as needed?

I don't see how that would work since `MethodInfo.Invoke` must necessarily box, which won't work for refs and ref-likes. There was a thread in the csharplang discussion about a language-only solution to store it only for GC purposes for static functions, though.

**Comment by TahirAhmadov** [comment]
> If struct delegates become a thing, they will either be structural or nominal, and likely never revisited again.

I'm not sure why it's necessarily an either-or. One possible approach would be, underlying system is nominal, and BCL gets `ValueAction`/etc. types, and C# syntax is added which compiles to those types; and even the conversion I proposed above can be made implicit between those new `ValueAction` types - if you really like that approach, fine, you can use it and it works like you expect. 
I just can't imagine how we can push this feature forward if it doesn't even allow for nominal struct delegate types. The type alias approach then becomes just the same thing, but 1) without type safety, 2) introduces the complication of variadic type arguments, and 3) a huge change from how all delegates used to work until now.
PS. I, for one, will most likely not even use it without nominal types. As I've written before, 95% of my delegates are nominal, and 99% of my structs are, as well. These "anonymous" tuple and delegate types never work in the mid term, let alone long term; unreadable, easy to make a mistake, etc.

> I don't see how that would work since `MethodInfo.Invoke` must necessarily box, which won't work for refs and ref-likes.

I see what you mean, we can't even represent the call in IL, which precludes JIT optimization to begin with. Well, it was an idea.

**Comment by timcassell** [comment]
`ValueAction` nominal type wouldn't become structural, the same way existing `Action` is not structural. For aliased type safety, I think the C# "roles" proposal would get you what you want.

**Comment by TahirAhmadov** [comment]
Even if so, how does that negatively impact that usage scenario? If you get the same syntax and behavior, what does it matter how the underlying stuff is implemented?

**Comment by timcassell** [comment]
What same syntax and behavior? I can't use `Action<ref T, out U>`.

**Comment by TahirAhmadov** [comment]
You can't today, but it can be made possible with new syntax, no? Fundamentally, `ref` is a type, and direction can be made an attribute.

I do think you should reassess your aversion to nominal delegate types. I find it hard to believe that copying and pasting stuff like `Action<ref int, string, out bool, in Guid>` is preferable to a simple identifier.

**Comment by timcassell** [comment]
I prefer to not have extra boilerplate of delegate creation, and I like to see at the usage place what the signature of the delegate is without having to jump to the delegate declaration. It's also immediately clear to me when a parameter is a delegate as opposed to some other type. And as a library author, I'd rather not pollute the amount of delegate types in the app.

**Comment by TahirAhmadov** [comment]
> I prefer to not have extra boilerplate of delegate creation,

The moment the same delegate type is referred to more than once, the cost shifts: you get the boilerplate of delegate usage. 
And I think when it comes to simple "pass a delegate to one method and it gets used in that method" scenario, more often than not it's a simple delegate like `Func<,>` or similar, and those do exist (for better or for worse). It's the complex ones with `ref/out` etc. stuff that I don't understand the need for "saving boilerplate of creation".
Even if you have a method which takes a delegate and it's called in that method (meaning, it's not passed/stored anywhere), it's more difficult to read together with other parameters (if you have them): literally yesterday I was reading my own code from a year ago where I used an anonymous delegate, and it all got blurred in my eyes, which identifier is part of the delegate generics, which one is a parameter type, etc. 
PS. Needless to say, I immediately got rid of all that stuff and declare `private` nested delegate and record struct types (for tuples); took me all of 2 minutes, and then in 10 seconds, I made sense of what my OWN code did; I was struggling to do so with the anonymous stuff.

>  I like to see at the usage place what the signature of the delegate is

That same logic can be extended to most DTOs - you don't want to jump to `Person` class to see it has a first name and a last name, why not just pass everything by tuple? If we're doing away with nominal delegate types altogether, that means you want to see all 8, 18, or however parameters it has - how is this different from a data object of some sort with N properties?

> It's also immediately clear to me when a parameter is a delegate as opposed to some other type.

Meh, it's easy enough to append the word Delegate to the type name.

> And as a library author, I'd rather not pollute the amount of delegate types in the app.

What's the problem with having more types in your library? 

Bottom line is, I don't mind you writing code with more "anonymous" types, like these "structural" delegates, but you are advocating for taking away the ability of others to create nominal types of delegates. I don't think I want to convince you to change your ways, but I do hope you grant me and others that we want to do it our way.

**Comment by timcassell** [comment]
> The moment the same delegate type is referred to more than once, the cost shifts: you get the boilerplate of delegate usage. And I think when it comes to simple "pass a delegate to one method and it gets used in that method" scenario, more often than not it's a simple delegate like `Func<,>` or similar, and those do exist (for better or for worse). It's the complex ones with `ref/out` etc. stuff that I don't understand the need for "saving boilerplate of creation". Even if you have a method which takes a delegate and it's called in that method (meaning, it's not passed/stored anywhere), it's more difficult to read together with other parameters (if you have them): literally yesterday I was reading my own code from a year ago where I used an anonymous delegate, and it all got blurred in my eyes, which identifier is part of the delegate generics, which one is a parameter type, etc.

When that happens, I just separate the parameters to new lines. It's very easy to read.

> That same logic can be extended to most DTOs - you don't want to jump to `Person` class to see it has a first name and a last name, why not just pass everything by tuple? If we're doing away with nominal delegate types altogether, that means you want to see all 8, 18, or however parameters it has - how is this different from a data object of some sort with N properties?

Exactly, and I love tuples for that reason also. I use tuples all the time for e.g. passing captured data to a delegate to avoid closure allocations. Or intermediate transforms in Linq queries.

> What's the problem with having more types in your library?

It's not my library, it's all the libraries and user assembly combining all of their declared delegates with potentially duplicate signatures. If it's structural, it's all runtime and there are no duplicates.

> Bottom line is, I don't mind you writing code with more "anonymous" types, like these "structural" delegates, but you are advocating for taking away the ability of others to create nominal types of delegates. I don't think I want to convince you to change your ways, but I do hope you grant me and others that we want to do it our way.

I'm not advocating for removing the ability to do what you like. I just think there are other ways to achieve it besides nominal delegates (like type aliases and roles).

**Comment by TahirAhmadov** [comment]
> When that happens, I just separate the parameters to new lines. It's very easy to read.

Which is more work, put new lines, indent everything nicely, etc. Not really "less work".

> I use tuples all the time for e.g. passing captured data to a delegate to avoid closure allocations. Or intermediate transforms in Linq queries.

These are valid use cases - come to think of it, the only times I use them is to replace allocated anonymous types, and prevent closure allocations. I do think we could have better solutions for those; I created a C# discussion for an allocation-free closure for immediately called methods. The entire LINQ is overdue for optimization itself with struct enumerables, etc. 
However, I was referring to method return types, fields, etc. - in none of those scenarios, do tuples make the code easier to read and maintain. I experienced it myself - I quickly put together a tuple, come back a few months later, and I'm trying to remember what I did, Nominal types are self-documenting. It's not about writing code fast, rather about writing code which can be maintained and improved over time.

> It's not my library, it's all the libraries and user assembly combining all of their declared delegates with potentially duplicate signatures. If it's structural, it's all runtime and there are no duplicates.

If these delegate types are internal, then nobody sees them outside, so they don't pollute. And if they're public, they better have a name, plus documentation for what it's supposed to do. And we do NOT want unrelated delegates from various assemblies to be assignable to each other just because the signatures match. It's like saying, you have a `Student` with first name and last name, and `Teacher` with same fields, so variable of type "student" can implicitly get a value of type "teacher" (because there are no types, per se) and developer doesn't get any warning whatsoever. It's a pit of failure.

> I'm not advocating for removing the ability to do what you like. I just think there are other ways to achieve it besides nominal delegates (like type aliases and roles).

None of those give me the ability to have a type safe reference to a method.

**Comment by TahirAhmadov** [comment]
PS. I wish we could get a combination of anonymous types and tuples; meaning, a syntax which would create a strongly-typed inline-defined record struct, for the LINQ stuff. It's unlikely to happen at this point, but that's what I would do in a clean-slate C#.

**Comment by TahirAhmadov** [comment]
PPS. Even worse, you have a "student" with first and last names, and a "vehicle" with make and model. One wrong variable assignment, and you have Chevy Malibu trying to learn ancient Greek literature. 🤣 Sorry, but no. Just a bad idea, IMO.

**Comment by timcassell** [comment]
The roles feature absolutely would give you delegate type safety (a shame it got put on the back-burner during extension everything). One other thing that I think would help with the self-documenting problem is if C# supported named type arguments (I remember seeing a proposal for it, but I don't have the link on hand).

**Comment by TahirAhmadov** [comment]
So we have 2 options - expensive runtime changes to support variadic type arguments, and expensive runtime changes to support roles the way their proponents would like them to work, in order to make this feature structural first; or neither one of those costs in terms runtime of work, with nominal first approach, and structural being just syntax sugar on top. The function pointer problem needs to be solved in both cases. So why choose the vastly more expensive option? 
PS. The syntax for "structural" delegate syntax can already be added to existing delegates and some work has already been done in that direction (to my dismay). That's why there's no reason to necessarily couple the two concepts for value type delegates; and this is ignoring the whole debate over that structural approach even being a good pattern to follow. Basically, we have a pattern which is at best controversial, which can be added as a language feature, but you insist that it's baked in at great cost purely out of your aversion to nominal types. I fail to see the rationale behind this stance.

**Comment by TahirAhmadov** [comment]
Regarding `MethodInfo`, I understand `Invoke` calls cannot be generated in IL for ref types, but can the IL contain something to convert the underlying method reference to a pointer and call that like you're proposing, with the understanding that JIT will understand this "dirty" sequence and optimize it to a simple call in machine code? 
PS. I keep thinking in this direction because I have a suspicion that even this weird workaround is easier to accomplish than pointer lifetime tracking (but I may be wrong on this. )

**Comment by timcassell** [comment]
Delegates cannot be made structural on top of nominal types purely through language changes. It will require runtime changes either way.

Delegates are a fundamental type. The fact that they are nominal has always seemed strange to me. Function pointers are structural, this is pretty much aligning them. There's not much difference between a function pointer and a delegate except 1 carries state (and all the other historical baggage that's not part of this proposal).

Either way it's a lot of work. If the end result in syntax is the same, then really the only tradeoff is whether delegates declared in separate assemblies with matching signatures have the same underlying runtime type.

**Comment by timcassell** [comment]
If struct delegates go a similar route as class delegates, the runtime can implement it as it sees fit, using a method info or modifying the GC. I could imagine it using method info for the initial implementation, then later use direct function pointer for more performance.

**Comment by TahirAhmadov** [comment]
> Delegates cannot be made structural on top of nominal types purely through language changes. It will require runtime changes either way.

They won't be "structural" really, but surely we can devise syntax which uses new BCL types to achieve a similar effect?

> Delegates are a fundamental type. The fact that they are nominal has always seemed strange to me. Function pointers are structural, this is pretty much aligning them. There's not much difference between a function pointer and a delegate

The same can be said about any pointer, "a class reference is just a pointer which is tracked by GC", etc. But there is a difference, and a huge difference - safety. Just like .NET mandates that references point at objects, not just at memory, delegates are a type safe way to invoke a method, not just a pointer to `call` with machine code. I totally agree with you that the heap storage of them and multicast stuff is unnecessary overhead, but it doesn't suddenly make struct delegates and function pointers identical.

> Either way it's a lot of work. If the end result in syntax is the same, then really the only tradeoff is whether delegates declared in separate assemblies with matching signatures have the same underlying runtime type.

I disagree. I think the whole "we make a magical struct which can have unlimited number of type arguments" is a much bigger change, than creating another copy-pasta of 16 types in BCL to be the basis of "syntax" "structural" delegates on the language level.
